File: 200-Number-of-Islands/bfs.js
/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function (grid) {
    const rows = grid.length, cols = grid[0].length;

    const directions = [[-1, 0], [1, 0], [0, 1], [0, -1]];
    let count = 0;

    const bfs = (r, c) => {
        const queue = [[r, c]];

        while (queue.length > 0) {
            let [r, c] = queue.pop();
            grid[r][c] = "*";

            for (const [dr, dc] of directions) {
                const row = r + dr, col = c + dc;
                if (row < 0 || col < 0 || row === rows || col === cols || grid[row][col] !== "1") {
                    continue;
                }
                queue.push([row, col]);
            }
        }
    }

    for (let r = 0; r < rows; r++) {
        for (let c = 0; c < cols; c++) {
            if (grid[r][c] === "1") {
                count += 1;
                bfs(r, c);
            }
        }
    }

    return count;
};

File: 200-Number-of-Islands/bfs.py
from typing import List


class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        directions = [(-1, 0), (1, 0), (0, 1), (0, -1)]
        islands = 0
        rows, cols = len(grid), len(grid[0])

        def bfs(r, c):
            q = [(r, c)]

            while q:
                r, c = q.pop()
                grid[r][c] = "*"
                for dr, dc in directions:
                    row, col = r + dr, c + dc
                    if row < 0 or row == rows or col < 0 or col == cols or grid[row][col] != "1":
                        continue
                    q.append((row, col))

        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == "1":
                    islands += 1
                    bfs(r, c)

        return islands